core/src/main/cpp/gdn/projection_mesh/projection_mesh.h

```diff
@@ -0,0 +1,209 @@
+            external_texture = Object::cast_to<ExternalTexture>(*texture);
+        }
+        return external_texture;
+    }
+
+    inline void set_render_on_top(bool enable) {
+        if (this->render_on_top == enable) {
+            return;
+        }
+        this->render_on_top = enable;
+        if (shader_material_ref.is_valid() && shader_material_ref->get_shader().is_valid()) {
+            String shader_code = shader_material_ref->get_shader()->get_code();
+            shader_code += kDisableDepthTestRenderMode;
+            shader_material_ref->get_shader()->set_code(shader_code);
+        }
+        update_render_priority();
+    }
+
+    inline bool is_render_on_top() {
+        return this->render_on_top;
+    }
+
+    inline void set_gaze_tracking(bool gaze_tracking) {
+        if (this->gaze_tracking == gaze_tracking) {
+            return;
+        }
+        this->gaze_tracking = gaze_tracking;
+        update_render_priority();
+        if (shader_material_ref.is_valid()) {
+            shader_material_ref->set_shader_param(kGastEnableBillBoardParamName, gaze_tracking);
+        }
+    }
+
+    inline bool is_gaze_tracking() {
+        return gaze_tracking;
+    }
+
+    inline void update_render_priority() {
+        ShaderMaterial *shader_material = *shader_material_ref;
+        if (!shader_material) {
+            return;
+        }
+
+        int render_priority = 0;
+        if (render_on_top) {
+            render_priority++;
+            if (gaze_tracking) {
+                render_priority++;
+            }
+        }
+        shader_material->set_render_priority(render_priority);
+    }
+
+    inline void set_alpha(float alpha) {
+        if (this->alpha == alpha) {
+            return;
+        }
+        this->alpha = alpha;
+        shader_material_ref->set_shader_param(kGastNodeAlphaParamName, alpha);
+    }
+
+    inline float get_gradient_height_ratio() {
+        return gradient_height_ratio;
+    }
+
+    inline void set_gradient_height_ratio(float ratio) {
+        if (this->gradient_height_ratio == ratio) {
+            return;
+        }
+        this->gradient_height_ratio = std::min(1.0f, std::max(0.0f, ratio));
+        shader_material_ref->set_shader_param(kGastGradientHeightRatioParamName, gradient_height_ratio);
+    }
+
+    inline StereoMode get_stereo_mode() {
+        return stereo_mode;
+    }
+
+    inline void set_stereo_mode() {
+        // TODO: Implement
+    }
+
+    inline void reset_mesh() {
+        mesh_instance->set_mesh(Ref<Resource>());
+    }
+
+  virtual Vector2 get_relative_collision_point(Vector3 local_collision_point) {
+        return kInvalidCoordinate;
+    }
+
+ private:
+    ProjectionMeshType projection_mesh_type;
+    MeshInstance *mesh_instance;
+    Ref<ShaderMaterial> shader_material_ref;
+    Ref<Shape> collision_shape_ref;
+    StereoMode stereo_mode;
+    bool render_on_top;
+    bool gaze_tracking;
+    float alpha;
+    float gradient_height_ratio;
+
+protected:
```

Comment:
The `ProjectionMesh` class should have the `generate_shader_code()` method as a protected method containing the common shader logic.
The derived classes can then override it and add their custom changes to the base shader.

Reply:
Done.